docs/proposals/scheduling/20220530-fine-grained-cpu-orchestration.md

```diff
@@ -0,0 +1,690 @@
+
+```go
+const (
+  FullPCPUsOnlyOption            string = "full-pcpus-only"
+  DistributeCPUsAcrossNUMAOption string = "distribute-cpus-across-numa"
+)
+
+type KubeletCPUManagerPolicy struct {
+  Policy  string            `json:"policy,omitempty"`
+  Options map[string]string `json:"options,omitempty"`
+}
+
+```
+
+- The annotation `node.koordinator.sh/cpu-topology` describes the detailed CPU topology. Fine-grained management mechanism needs more detailed CPU topology information. The scheme is defined as follows:
+
+```go
+type CPUTopology struct {
+  Detail []CPUInfo `json:"detail,omitempty"`
+}
+
+type CPUInfo struct {
+  CPUID        int32 `json:"cpu"`
+  CoreID       int32 `json:"core"`
+  NUMASocketID int32 `json:"numaSocket"`
+  NUMANodeID   int32 `json:"numaNode"`
+}
+```
+
+- annotation `node.koordinator.sh/pod-cpu-allocs` describes the CPUs allocated by Koordinator LSE/LSR and K8s Guaranteed Pods. The scheme corresponding to the annotation value is defined as follows:
+
+```go
+type PodCPUAlloc struct {
+  Namespace string    `json:"namespace,omitempty"`
+  Name      string    `json:"name,omitempty"`
+  UID       types.UID `json:"uid,omitempty"`
+  CPU       string    `json:"cpu,omitempty"`
+}
+
+type PodCPUAllocs []PodCPUAlloc
+```
+
+- The annotation `node.koordinator.sh/cpu-shared-pools` describes the CPU Shared Pool defined by Koordinator. The shared pool is mainly used by Koordinator LS Pods or K8s Burstable Pods. The scheme is defined as follows:
+
+```go
+type NUMACPUSharedPools []CPUSharedPool
+
+type CPUSharedPool struct {
+  NUMASocketID int32  `json:"numaSocket"`
+  NUMANodeID   int32  `json:"numaNode"`
+  CPUSet       string `json:"cpuset,omitempty"`
+}
+```
+The `CPUSet` field is Linux CPU list formatted string. For more details, please refer to [doc](http://man7.org/linux/man-pages/man7/cpuset.7.html#FORMATS).
+
+
+##### Create/Update NodeResourceTopology
+
+- koordlet is responsible for creating/updating NodeResourceTopology
+- It is recommended that koordlet obtain the CPU allocation information of the existing K8s Guaranteed Pod by parsing the CPU state checkpoint file. Or obtain this information through the CRI interface and gRPC provided by kubelet.
+- When the CPU of the Pod is allocated by koord-scheduler, replace the CPUs in the kubelet state checkpoint file.
+- It is recommended that koordlet obtain the CPU manager policy and options from [kubeletConfiguration](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/).
+
+##### Example
+
+A complete NodeResourceTopology example:
+
+```yaml
+apiVersion: topology.node.k8s.io/v1alpha1
+kind: NodeResourceTopology
+metadata:
+  annotations:
+    kubelet.koordinator.sh/cpu-manger-policy: |-
+      {
+        "policy": "static",
+        "options": {
+          "full-pcpus-only": "true",
+          "distribute-cpus-across-numa": "true"
+        }
+      }
+    node.koordinator.sh/cpu-topology: |-
```

Comment:
Is this a duplicate of the following zone info?

Reply:
Currently NodeResourceTopology does not have specific CPU ID. However, we can also consider using the Zone.Attributes field to record which CPU IDs are in a NUMA Node.

Reply:
After consideration, keep the design in the proposal. The reason is that NodeResourceTopology does not have specific information such as logical CPU ID and physical core ID. If this information is added to Attribute, it will be bloated. I think in the future, it may be possible to cooperate with upstream to expand new fields to solve.